pkg/destroy/azure/azure.go

```diff
@@ -0,0 +1,247 @@
+	logger = logger.WithField("dns zone", zoneName)
+	sharedZone, err := getSharedDNSZone(ctx, dnsClient, zoneName)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to find shared zone for %s", zoneName)
+		if errors.Cause(err) == errNoMatchingSharedZone {
+			logger.WithField("private zone", zoneName).Debugf(err.Error())
+			return nil
+		}
+		return err
+	}
+	logger.Debugf("found shared zone %s", sharedZone.Name)
+	allPrivateRecords := sets.NewString()
+	for recordPages, err := recordsClient.ListByDNSZone(ctx, zoneGroup, zoneName, to.Int32Ptr(100), ""); recordPages.NotDone(); err = recordPages.NextWithContext(ctx) {
+		if err != nil {
+			return err
+		}
+		for _, record := range recordPages.Values() {
+			if t := toRecordType(to.String(record.Type)); t == dns.SOA || t == dns.NS {
+				continue
+			}
+			allPrivateRecords.Insert(fmt.Sprintf("%s.%s", to.String(record.Name), zoneName))
+		}
+	}
+
+	for recordPages, err := recordsClient.ListByDNSZone(ctx, sharedZone.Group, sharedZone.Name, to.Int32Ptr(100), ""); recordPages.NotDone(); err = recordPages.NextWithContext(ctx) {
+		if err != nil {
+			return err
+		}
+		for _, record := range recordPages.Values() {
+			if allPrivateRecords.Has(fmt.Sprintf("%s.%s", to.String(record.Name), sharedZone.Name)) {
+				resp, err := recordsClient.Delete(ctx, sharedZone.Group, sharedZone.Name, to.String(record.Name), toRecordType(to.String(record.Type)), "")
+				if err != nil {
+					if wasNotFound(resp.Response) {
+						logger.WithField("record", to.String(record.Name)).Debug("already deleted")
+						continue
+					}
+					return errors.Wrapf(err, "failed to delete record %s", to.String(record.Name))
+				}
+				logger.WithField("record", to.String(record.Name)).Info("deleted")
+			}
+		}
+	}
+	return nil
+}
+
+// getSharedDNSZone returns the closest parent public dns zone for privZoneName
+func getSharedDNSZone(ctx context.Context, client dns.ZonesClient, privZoneName string) (dnsZone, error) {
+	domain := privZoneName
+	parents := sets.NewString(domain)
+	for {
+		idx := strings.Index(domain, ".")
+		if idx == -1 {
+			break
+		}
+		if len(domain[idx+1:]) > 0 {
+			parents.Insert(domain[idx+1:])
+		}
+		domain = domain[idx+1:]
+	}
+
+	allPublicZones := []dnsZone{}
+	for zonesPage, err := client.List(ctx, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
+		if err != nil {
+			return dnsZone{}, err
+		}
+		for _, zone := range zonesPage.Values() {
+			if zone.ZoneType == dns.Public && parents.Has(to.String(zone.Name)) {
+				allPublicZones = append(allPublicZones, dnsZone{Name: to.String(zone.Name), ID: to.String(zone.ID), Group: groupFromID(to.String(zone.ID)), Public: true})
+				continue
+			}
+		}
+	}
+	sort.Slice(allPublicZones, func(i, j int) bool { return len(allPublicZones[i].Name) > len(allPublicZones[j].Name) })
+	if len(allPublicZones) > 0 {
+		return allPublicZones[0], nil
```

Comment:
Do we want to return all of these and have the destroyer remove matching records from each?  Nipping the #1769-style ambiguity in the bud for this code.

Reply:
done with https://github.com/openshift/installer/pull/1771/commits/3d25470009039e20339d645c8a4454559406e914 -> https://github.com/openshift/installer/pull/1771/commits/27da87f528d10f6dd1d7678e7ec859365c1df300